### PR TITLE
client/core,eth: Lock funds for refund

### LINKED ISSUE #1466: client/eth: Funds locked to be used for refund are unlocked too early
The max possible funds to be used for both initiating swaps and refunds are locked in `FundOrder`, but after the last swap is initiated, all of these funds are unlocked. The funds reserved for refunds should only be unlocked when it is certain there will no longer be any more refunds.

### PR DESCRIPTION
**Pull request by martonp**
Previously, the funds needed for refunds were locked in `FundOrder` together with the funds needed for initiations, but they were unlocked when `Swap` was called with `LockChange = false`.  This diff applies the same logic we use to reserve funds for redemptions to refunds.

Prior to this PR, redemptions were being reserved using the server's max fee rate, but the redemption transaction was being submitted with the current fee rate estimate. This PR updates this to use the server's max fee rate when submitting the transaction as well. There is no reason to use a lower rate when submitting the transaction, as ETH uses dynamic transaction fees, and the entire fee will not be used anyways. The server's max fee rate is also used for refunds.

- `client/asset`: Rename `AssetRedeemer` interface to `AssetLocker`, and add functions for refunds with the same functionality as already exist for redemptions.
- `client/eth`: Refactor fund locking to keep track of funds locked for initiation, refunds, and redemptions separately. Also, make ETH a `FeeRater`.
- `client/core`: Keep track of refund reserves the same way as they are tracked for redemptions. In addition to unlocking when the deal is revoked or cancelled, funds reserved for refunds are also unlocked when either the counterparty or the current user does a redemption.

Closes #1466

### CODE REVIEW

File: client/asset/eth/eth.go
Comment (martonp) [reply client/asset/eth/eth.go:1184]: I'll just log the error in the caller. To log the error in unlock I'd have to give the logger to `FundReserves` which is kind of strange.

File: client/core/core.go
Comment (martonp) [reply client/core/core.go:5097]: These things are checked in `lockRefundFraction` and `lockRedemptionFraction`.

**Comment by chappjc** [comment]
Minor conflict by proximity to FeeRater changes in core_test.go.  Not holding up my review though, that's coming shortly.

File: client/core/core.go
Comment (JoeGruffins) [reply client/core/core.go:5097]: OK! I see it now. the `t.accountRefunder()` and `t.accountRedeemer()` checks. Thanks.

**Review by JoeGruffins — APPROVED** [review]
Working well.

**Review by chappjc — COMMENTED** [review]
Some quick comments.  I think we need to have a discussion in matrix about making the eth client a FeeRate with the formula we've settled on, and then perhaps capping the computed rate by the user's fee rate limit.  related: https://github.com/decred/dcrdex/pull/1485

File: client/db/types.go
Comment (martonp) [reply client/db/types.go:270]: These two fields actually never get updated. They are the initial amount reserved by the wallet for redemptions and refunds. Only the `redemptionLocked` and `refundLocked` fields in `trackedTrade` get updated.

Comment (chappjc) [reply client/db/types.go:270]: What happens if the app is restarted with a live order and the order data reloaded from DB for a partially filled order?  Wouldn't these fields need to persist the values in the `trackedTrade`?

Comment (martonp) [reply client/db/types.go:270]: This is handled in `resumeTrades`.  `redemptionLocked` and `refundLocked` are updated based on how many redemptions/refunds are still possibly remaining.

Comment (chappjc) [reply client/db/types.go:270]: Gotcha.  I see `lockRedemptionFraction` now

File: client/asset/eth/eth.go
Comment (chappjc) [reply client/asset/eth/eth.go:923]: Never mind my previous comment about FeeRate for refund or redeem txns.  On matrix we agreed it's safest and simplest to use the MaxFeeRate as the max gas price for the (dynamic) refund and redeem txns because nothing is gained by using a lower estimate.

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:4209]: Works for me.  Only additional thought about this matter is that the rate given to `ReserveNRedemptions` and `ReserveNRefunds` could be the larger of `MaxFeeRate` and the rate from `FeeRate` (if also a `FeeRater`).  That would be pretty screwy if the server had a rate set that low, so whatever, let's not change anything now.  This PR looks good to me.

File: client/core/trade.go
Comment (chappjc) [inline client/core/trade.go:170]: Unimportant, but the preferred syntax would be `redemptionReserves, refundReserves uint64` without the redundant type.

File: client/db/bolt/db.go
Comment (chappjc) [inline client/db/bolt/db.go:98]: For the record, I'm on board with breaking compatibility with the last dev version of the DB encoding.  Not worth an upgrade esp. considering the state of ETH (not used in the wild).

File: client/core/core.go
Comment (martonp) [reply client/core/core.go:4209]: If the swap cannot be initiated, then it doesn't matter if too little is reserved for refunds and redemptions.

Comment (chappjc) [reply client/core/core.go:4209]: The swap tx could still be mined with a fee less than what FeeRate would recommend.

File: client/asset/eth/eth.go
Comment (chappjc) [inline client/asset/eth/eth.go:257]: I'm tempted to say panic, remove the error and logging in `unlock`'s caller.
With an enum of `fundReserveType`s, it takes a special type of programmer error to pass an invalid value.

**Review by chappjc — APPROVED** [review]
LGTM.  Kinda too bad that the `unlock` method has an error return though.

**Comment by chappjc** [comment]
Ready to merge unless @buck54321 is reviewing.

File: client/core/trade.go
Comment (buck54321) [inline client/core/trade.go:255]: This is a change in behavior, even if the outcome may be identical. Before, if the requested reservers were `> redemptionReserves`, an error was logged and that's it. Now, an error is logged, but we still call `ReReserveRedemption` with an argument of `0`, and add `0` to the `redemptionLocked`. I don't expect that this would result in any issues, but calling `ReReserveRedemption` with `0` is questionable behavior.

File: client/asset/interface.go
Comment (buck54321) [inline client/asset/interface.go:375]: This method would be unneeded if we lock up the refund reserves in `FundOrder`.

File: client/asset/eth/eth.go
Comment (chappjc) [reply client/asset/eth/eth.go:262]: That's a good point.  Could even embed the mutex in that struct

File: client/asset/eth/eth.go
Comment (martonp) [reply client/asset/eth/eth.go:657]: I was thinking about that too, but I did it this way because if the refund reserves were locked here, then `FundOrder` would need to return the amount locked for refunds separately. I didn't want this to have to be done in non-account based coins.

Comment (buck54321) [reply client/asset/eth/eth.go:657]: I see what you mean. I might propose a change later, but this is good for now.

**Review by buck54321 — CHANGES_REQUESTED** [review]
A few last things.

File: client/core/trade.go
Comment (martonp) [reply client/core/trade.go:703]: You mean `ImmediateTiF` right?

Comment (buck54321) [reply client/core/trade.go:703]: Yes. Sorry.

**Review by buck54321 — APPROVED** [review]

**Pull request merged**
